contracts/Remittance.sol

```diff
@@ -1,16 +1,118 @@
 
-    using SafeMath for uint;
+    using SafeMath for uint256;
 
+    mapping(address => uint) public balances;
 
+    mapping(bytes32 => RemittanceInstance) public remittances;
+
+    struct RemittanceInstance {
+        address funder;
+        address broker;
+        uint amount;
+        bool fundsReleased;
+    }
+
+    event Deposit(
+        address indexed depositor,
+        uint amount
+    );
+
+    event RemittanceCreated(
+        bytes32 indexed hash,
+        address indexed funder,
+        address indexed broker,
+        uint amount
+    );
+
+    event RemittanceFundsReleased(
+        bytes32 indexed hash,
+        address indexed broker,
+        uint amount
+    );
+
+    event WithDraw(
+        address indexed withdrawer,
+        uint amount
+    );
+
+    // Generic deposit function
+    function deposit() external payable whenNotPaused {
+        require(msg.value > 0, "The value must be greater than 0");
+        balances[msg.sender] = balances[msg.sender].add(msg.value);
+
+        emit Deposit(msg.sender, msg.value);
+    }
+
+    // The funder can create a remittance
+    function create(bytes32 hash, address broker, uint256 amount)
+        external
+        whenNotPaused
+        returns(bool success)
+    {
+        require(broker != msg.sender, "The caller cannot be the broker");
+        require(amount > 0, "The amount must be greater than 0");
+
+        uint256 funderBalance = balances[msg.sender];
+        require(funderBalance >= amount, "There are insufficient funds in the funder's account to create this remittance");
+
+        // Subtract from Funder balance
+        balances[msg.sender] = SafeMath.sub(funderBalance, amount);
+
+        RemittanceInstance storage remittanceInstance = remittances[hash];
+
+        remittanceInstance.funder = msg.sender;
+        remittanceInstance.broker = broker;
+        remittanceInstance.amount = amount;
+        remittanceInstance.fundsReleased = false;
+
+        emit RemittanceCreated(hash, msg.sender, broker, amount);
+
+        return true;
+    }
+
+    // The broker takes concatenates their password with the recipient's password to release the funds
+    function release(string memory concatenatedPassword) public whenNotPaused returns(bool success) {
+
+        require(bytes(concatenatedPassword).length > 0, "The concatenated password cannot be empty");
+        bytes32 hash = keccak256(abi.encodePacked(concatenatedPassword));
+
+        //Retrieve remittance
+        RemittanceInstance storage remittanceInstance = remittances[hash];
+        uint amount = remittanceInstance.amount;
+        address broker = remittanceInstance.broker;
+
+        require(remittanceInstance.fundsReleased == false, "The funds have already been released");
+        require(msg.sender == broker, "Only the broker can release funds for this remittance");
+
+        emit RemittanceFundsReleased(hash, msg.sender, amount);
+        remittanceInstance.fundsReleased = true;
```

Comment:
Can you tell me how useful this `.fundsReleased` is?

Reply:
Also, you know, you can get some gas back when you set the storage back to `0`. Any idea what could go to `0`?

Reply:
I did not know that. 

My reasoning is that I want the blockchain to store a record that the transaction completed successfully. Based on what you say. I think can make the following changes:

- Delete 'fundsReleased'
- Rename 'amount' to 'fundsOwed'
- Set the fundsOwed to 0 after the release of funds

Presumably the contract receives gas if a redundant struct record from a mapping hash table is removed? So I need to think about how I can prune the storage and keeping it as small as possible. I understand that in this contract I need to keep each struct instance as a permanent record unless the money is never claimed and then it could be removed.

Reply:
Yes, your 3 actions, when done together, do not change the safety or functioning of the contract, and do it for a cheaper gas cost.
When you zero out a field, it is the inner transaction that gets the gas back, before it bubbles up the call stack. So it is not the contract per se that collects it.
And yes, as long as money is owed, it cannot be zeroed out, that would be akin to burning other people's money.

Reply:
> And yes, as long as money is owed, it cannot be zeroed out, that would be akin to burning other people's money.

In an Escrow situation then I understand the need to have a time limit so that the funds can be returned if the broker never collects the money held in Escrow.

Reply:
Yes, let's introduce a time limit.

Reply:
> Yes, let's introduce a time limit.

Will do.